```
/usr/bin/ld: CMakeFiles/makefixed.dir/__/inftrees.c.o: in function `zng_inflate_table':
/home/runner/work/zlib-ng/zlib-ng/inftrees.c:286:(.text+0x404): undefined reference to `z_error'
/usr/bin/ld: ../makefixed: hidden symbol `z_error' isn't defined
```

--- a/inftrees.c
+++ b/inftrees.c
@@ -4,6 +4,7 @@
  */
 
 #include "zbuild.h"
+#include "zmemory.h"
 #include "zutil.h"
 #include "inftrees.h"
 #include "inflate_p.h"
@@ -281,6 +282,9 @@ int Z_INTERNAL zng_inflate_table(codetype type, uint16_t *lens, unsigned codes,
        in the rest of the decoding tables with invalid code markers.
      */
 
+    /* The packed table fill below reads and writes each entry as a single uint32_t. */
+    Assert(sizeof(code) == 4, "code entry must be exactly 4 bytes");
+
     /* set up for code type */
     switch (type) {
     case CODES:
@@ -336,10 +340,25 @@ int Z_INTERNAL zng_inflate_table(codetype type, uint16_t *lens, unsigned codes,
         incr = 1U << (len - drop);
         fill = 1U << curr;
         min = fill;                 /* save offset to next table */
-        do {
-            fill -= incr;
-            next[(huff >> drop) + fill] = here;
-        } while (fill != 0);
+        {
+            /* `here` is a 4-byte struct that the compiler writes one field at a time.
+               Read it as a packed uint32_t so each slot is filled by a single store. */
+            uint32_t here_u32 = zng_memread_4(&here);
+            code *base_ptr = next + (huff >> drop);
+            unsigned step4 = incr << 2;
+            /* Unroll by four so the slot addresses are independent. */
+            while (fill >= step4) {
+                zng_memwrite_4(&base_ptr[fill - 1 * incr], here_u32);
+                zng_memwrite_4(&base_ptr[fill - 2 * incr], here_u32);
+                zng_memwrite_4(&base_ptr[fill - 3 * incr], here_u32);
+                zng_memwrite_4(&base_ptr[fill - 4 * incr], here_u32);
+                fill -= step4;
+            }
+            while (fill != 0) {
+                fill -= incr;
+                zng_memwrite_4(&base_ptr[fill], here_u32);
+            }
+        }
 
         /* backwards increment the len-bit code huff */
         rhuff = (uint16_t)(rhuff + (0x8000u >> (len - 1)));

--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -25,7 +25,7 @@ add_executable(infcover infcover.c)
 configure_test_executable(infcover)
 target_link_libraries(infcover zlib-ng)
 if(NOT DEFINED BUILD_SHARED_LIBS OR BUILD_SHARED_LIBS)
-    target_sources(infcover PRIVATE ${PROJECT_SOURCE_DIR}/inftrees.c)
+    target_sources(infcover PRIVATE ${PROJECT_SOURCE_DIR}/inftrees.c ${PROJECT_SOURCE_DIR}/zutil.c)
 endif()
 add_test(NAME infcover COMMAND ${CMAKE_CROSSCOMPILING_EMULATOR} $<TARGET_FILE:infcover>)
 

--- a/utils/CMakeLists.txt
+++ b/utils/CMakeLists.txt
@@ -31,7 +31,7 @@ endif()
 set(MINIDEFLATE_COMMAND ${CMAKE_CROSSCOMPILING_EMULATOR} $<TARGET_FILE:minideflate> PARENT_SCOPE)
 
 # makefixed
-add_executable(makefixed makefixed.c ${PROJECT_SOURCE_DIR}/inftrees.c)
+add_executable(makefixed makefixed.c ${PROJECT_SOURCE_DIR}/inftrees.c ${PROJECT_SOURCE_DIR}/zutil.c)
 configure_util_executable(makefixed)
 set(MAKEFIXED_COMMAND ${CMAKE_CROSSCOMPILING_EMULATOR} $<TARGET_FILE:makefixed> PARENT_SCOPE)
 

